src/OpenTelemetry.Collector.Azure/AzureSdkCollector.cs

```diff
@@ -0,0 +1,150 @@
+﻿// <copyright file="AzureSdkCollector.cs" company="OpenTelemetry Authors">
+// Copyright 2018, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+namespace OpenTelemetry.Collectors.Azure
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using OpenTelemetry.Context;
+    using OpenTelemetry.Trace;
+
+    public class AzureSdkCollector : IDisposable, IObserver<DiagnosticListener>, IObserver<KeyValuePair<string, object>>
+    {
+        private readonly ConcurrentDictionary<Activity, IScope> scopes = new ConcurrentDictionary<Activity, IScope>(new ActivityReferenceEqualityComparer());
```

Comment:
IScope already should know about Activity (through ConditionalWeakTable :( ), we should avoid concurrent dictionary (coz we now need to think about cleaning it up, perf costs ,etc)

Reply:
Who creates this relationship between activity and the span?

Reply:
tracer.WithSpan(span)

Reply:
Ugh, things are complicated :)

Reply:
But how do I get the span back to dispose it?

Reply:
1. you don't need to create scope
2. When you need to stop span - do Tracer.CurrentSpan and end it
3. Azure SDK will stop the Activity

Reply:
wait a sec, let me check everything :)

Reply:
> When you need to stop span - do Tracer.CurrentSpan and end it

Tracer.CurrentSpan might not be the right one.

Reply:
so the idea is that Activity.Current is the source of truth: if you stop Activity it should be current. Otherwise, we are in trouble. From Azure SDK, I believe it is guaranteed that it always stops the current activity.
 
There is a 1:1 relation between Span and Activity - and tracer.CurrentSpan figures out the current span from the current Activity. 

The thing that attaches span to Activity is a tracer.WithSpan - so you kina have to call it now. But you get the scope back and you have no way of retaining this scope between start and stop - this is the problem. And you have no way to dispose the scope (unless you do concurrent dict).

I.e. basically scopes do not work well with libraries instrumented with Activities and we shouldn't be using them (they still work good in user code). All current auto-collectors have this issue now that they drop scope and never dispose it.

This makes me think that when span is created for Activity tracked by the library,  we should connect Activity to Span without creating scope and without tracer.WithSpan.

I'll send a PR for this and update existing auto-collectors to work with this pattern later today.

Reply:
So I've tried to get rid of scopes for auto-collected Activities in #162 and failed - something needs to attach span to current activity and unfortunately it returns scope.  But there is no reasonable way to dispose it.

I'll work more on this: https://github.com/open-telemetry/opentelemetry-dotnet/issues/163
For now, I suggest calling `trace.WithSpan` but not disposing the scope - we'll fix it everywhere in #163.

Reply:
How would I set a correct span length without disposing the scope?

Reply:
you'd end the current span, check out https://github.com/open-telemetry/opentelemetry-dotnet/blob/08599b639300ae7428159a7a6ff61dc5b6208378/src/OpenTelemetry.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs#L95

Reply:
Done